raysect/core/math/function/float/function1d/interpolate.pyx

```diff
@@ -0,0 +1,538 @@
+#        contributors may be used to endorse or promote products derived from
+#        this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+
+"""
+Interpolation functions for float.Function1D
+
+Interpolators are accessed through interface class Interpolate1D, which
+"""
+
+import numpy as np
+cimport cython
+from raysect.core.math.cython.interpolation.linear cimport linear1d
+from raysect.core.math.cython.interpolation.cubic cimport calc_coefficients_1d, evaluate_cubic_1d
+from raysect.core.math.cython.utility cimport find_index, lerp
+
+DEF INT_LINEAR = 0
+DEF INT_CUBIC = 1
+DEF INT_CUBIC_CONSTRAINED = 2
+
+_INTERPOLATION_TYPES = {
+    'linear': INT_LINEAR,
+    'cubic': INT_CUBIC,
+    'cubic_constrained': INT_CUBIC_CONSTRAINED
+}
+
+DEF EXT_NONE = 0
+DEF EXT_NEAREST = 1
+DEF EXT_LINEAR = 2
+DEF EXT_QUADRATIC = 3
+
+_EXTRAPOLATION_TYPES = {
+    'none': EXT_NONE,
+    'nearest': EXT_NEAREST,
+    'linear': EXT_LINEAR,
+    'quadratic': EXT_QUADRATIC
+}
+
+cdef class Interpolate1D(Function1D):
+    """
+    Interface class for Function1D interpolators
+
+    :param object x: 1D array-like object of real values.
+    :param object f: 1D array-like object of real values.
+    :param str interpolation_type: Type of interpolation to use ('linear', 'cubic', 'cubic_constrained')
+    :param str extrapolation_type: Type of extrapolation to use ('none', 'nearest', 'linear', 'quadratic')
+    """
+    def __init__(self, object x, object f, str interpolation_type,
+                 str extrapolation_type, double extrapolation_range):
+
+        interpolation_type = interpolation_type.lower()
+        if  interpolation_type not in _INTERPOLATION_TYPES:
+            raise ValueError(f'Interpolation type {interpolation_type} not found.')
+
+        cdef int interpolator_idx = _INTERPOLATION_TYPES.get(interpolation_type)
+
+        extrapolation_type = extrapolation_type.lower()
+        if extrapolation_type not in _EXTRAPOLATION_TYPES:
+            raise ValueError(f'Extrapolation type {extrapolation_type} not found.')
+
+        cdef int extrapolator_idx = _EXTRAPOLATION_TYPES.get(extrapolation_type)
+
+        self.x = np.array(x, dtype=np.float64)
```

Comment:
Allowing direct access to the data  means the user can modify the array which breaks the interpolation cache. For simplicity we assume that once an interpolator is instanced, the data it contains is immutable. A read-only numpy array should be returned (see the numpy api).